include/deal.II/lac/block_direct_solver.h

```diff
@@ -0,0 +1,407 @@
+    }
+
+  AssertDimension(searching_order.size(), n_mpi_process);
+
+  Assert(0 != n_blocks, ExcEmptyObject());
+  dof_shifts.resize(n_mpi_process);
+  for (unsigned int p = 0; p < n_mpi_process; ++p)
+    dof_shifts[p].resize(n_blocks);
+
+
+  std::vector<types::global_dof_index> starts(n_mpi_process, 0);
+  types::global_dof_index              n_dofs = 0;
+
+
+
+  for (unsigned int i = 1; i < n_mpi_process; ++i)
+    {
+      starts[i] = starts[i - 1];
+
+      for (unsigned int b = 0; b < n_blocks; ++b)
+        {
+          starts[i] += owned_partitioning[b][i - 1].n_elements();
+        }
+    }
+
+
+  for (unsigned int b = 0; b < n_blocks; ++b)
+    n_dofs += owned_partitioning[b][this_mpi_process].size();
+
+
+  for (unsigned int p = 0; p < n_mpi_process; ++p)
+    {
+      dof_shifts[p][0] =
+        starts[p] - owned_partitioning[0][p].nth_index_in_set(0);
+      for (unsigned int b = 1; b < n_blocks; ++b)
+        dof_shifts[p][b] = owned_partitioning[b - 1][p].nth_index_in_set(0) +
+                           dof_shifts[p][b - 1] +
+                           owned_partitioning[b - 1][p].n_elements() -
+                           owned_partitioning[b][p].nth_index_in_set(0);
+    }
+
+  shifted_owned.set_size(n_dofs);
+
+  const unsigned int begin_range =
+    dof_shifts[this_mpi_process][0] + owned_dofs[0].nth_index_in_set(0);
+  const unsigned int end_range = owned_dofs[n_blocks - 1].nth_index_in_set(0) +
+                                 dof_shifts[this_mpi_process][n_blocks - 1] +
+                                 owned_dofs[n_blocks - 1].n_elements();
+
+  shifted_owned.add_range(begin_range, end_range);
+
+
+  n_local_dofs = 0;
+  for (unsigned int b = 0; b < n_blocks; ++b)
+    n_local_dofs += owned_dofs[b].n_elements();
+
+  AssertDimension(n_local_dofs, shifted_owned.n_elements());
+
+  Assert(shifted_owned.is_ascending_and_one_to_one(mpi_comm),
+         ExcMessage(
+           "The shifted dofs are not valid index set (internal error)"));
+
+
+  first_block_index.resize(n_blocks);
+  last_block_index.resize(n_blocks);
+  for (unsigned int b = 0; b < n_blocks; ++b)
+    {
+      first_block_index[b] =
+        dof_shifts[this_mpi_process][b] + owned_dofs[b].nth_index_in_set(0);
+      last_block_index[b] =
+        first_block_index[b] + owned_dofs[b].n_elements() - 1;
+    }
+
+  AssertDimension(last_block_index[n_blocks - 1],
+                  first_block_index[0] + n_local_dofs - 1);
+}
+
+
+
+template <typename LevelNumber>
+template <class Operator>
+void
+BlockDirectSolver<LevelNumber>::initialize(const Operator &    vmult_operator,
+                                           const unsigned int &lvl)
+{
+  matrix = std::make_unique<TrilinosWrappers::SparseMatrix>(shifted_owned,
+                                                            shifted_owned,
+                                                            mpi_comm,
+                                                            100);
```

Comment:
Magic number?

Reply:
Yup. Magic.
The truth is, this was implemented to handle Stokes 2D coarse problem and since   compress in  trilinos matrix removes unneeded elements I've putted a very rough approximate. I should change it  and  move it to  AdditionalData structure.